src/main/java/spoon/reflect/visitor/PrinterTokenWriter.java

```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright (C) 2006-2017 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.reflect.visitor;
+
+import spoon.reflect.visitor.printer.TokenType;
+
+/**
+ * The token based and context aware printer.
+ */
+public interface PrinterTokenWriter {
+	/**
+	 * Writes one token of printed java source code to the registered {@link PrinterHelper}
+	 * @param type The type of the token
+	 * @param token the value of the token. The value type depends on TokenType. See {@link TokenType#getTokenClass()}
+	 */
+	void writeToken(TokenType type, Object token);
+
+	/**
+	 * @return next PrinterTokenWriter in the chain
+	 */
+	PrinterTokenWriter getNext();
+
+	/**
+	 * sets next PrinterTokenWriter in the chain
+	 */
+	void setNext(PrinterTokenWriter next);
```

Comment:
I don't understand the pattern here. Why do we need a chain? Does it have to be part of the API? Can this be encapsulated in DJPP or the helper?

Reply:
I fight with that problem too...
I think I need it to keep support of `Environment#isPreserveLineNumbers()`.
But may be my approach is really wrong, because feature "preserving of line numbers" is not conceptually compatible with TokenPrinterWriter concept, which should let client to format printer sources as client needs.

So may be we should somehow support two modes:
1) default - which supports isPreserveLineNumbers
2) client specific TokenPrinterWriter - which will ignore that setting.
WDYT?